src/issuance.rs

```diff
@@ -83,26 +83,37 @@ impl IssueAction {
         self.finalize
     }
 
-    /// Return the `AssetBase` if the provided `ik` is used to derive the `asset_id` for **all** internal notes.
+    /// Return the (`AssetBase`, `ValueSum`) tuple if the provided `ik` is used to derive
+    /// the `asset_id` for **all** internal notes.
+    // FIXME: consider renaming this function as it not only validates assets in notes
+    // but also calculates the sum of the note values.
     fn are_note_asset_ids_derived_correctly(
         &self,
         ik: &IssuanceValidatingKey,
-    ) -> Result<AssetBase, Error> {
-        match self
-            .notes
-            .iter()
-            .try_fold(self.notes().head.asset(), |asset, &note| {
-                // Fail if not all note types are equal
-                note.asset()
-                    .eq(&asset)
-                    .then(|| asset)
-                    .ok_or(IssueActionIncorrectNoteType)
-            }) {
-            Ok(asset) => asset // check that the asset was properly derived.
-                .eq(&AssetBase::derive(ik, &self.asset_desc))
-                .then(|| asset)
-                .ok_or(IssueBundleIkMismatchNoteType),
-            Err(e) => Err(e),
+    ) -> Result<(AssetBase, ValueSum), Error> {
+        // Calculate the value of the asset as a sum of values of all its notes
+        // and ensure all note types are equal
+        let (asset, value_sum) = self.notes.iter().try_fold(
+            (self.notes().head.asset(), ValueSum::zero()),
+            |(asset, value_sum), &note| {
+                // Update asset value (fail if not all note types are equal
+                // or if overflow occured)
+                if note.asset().eq(&asset) {
+                    let updated_value_sum =
+                        (value_sum + i128::from(note.value())).ok_or(ValueSumOverflow)?;
+                    Ok((asset, updated_value_sum))
+                } else {
+                    Err(IssueActionIncorrectNoteType)
+                }
+            },
+        )?;
```

Comment:
Strive for a more compact code (and comments):
```rust
        let (asset, value_sum) = self.notes.iter().try_fold(
            (self.notes().head.asset(), ValueSum::zero()),
            |(asset, value_sum), &note| {
                // all assets should have the same asset
                note.asset().eq(&asset).then(|| ()).ok_or(IssueActionIncorrectNoteType)?;
                // the total amount should not overflow
                Ok((asset, (value_sum + note.value()).ok_or(ValueSumOverflow)?))
            },
        )?;
```

Reply:
Done.